src/java/org/apache/cassandra/locator/DynamicEndpointSnitchEMA.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.locator;
+
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.cassandra.gms.Gossiper;
+import org.apache.cassandra.metrics.ExponentialMovingAverage;
+import org.apache.cassandra.net.LatencyMeasurementType;
+
+
+/**
+ * A dynamic snitching implementation that uses Exponentially Decaying Histograms to prefer or
+ * de-prefer hosts
+ *
+ * This was the default implementation prior to Cassandra 4.0 and is being left as the default
+ * in 4.0
+ */
+public class DynamicEndpointSnitchEMA extends DynamicEndpointSnitch
+{
+    // A ~10 sample EMA
+    private static final double EMA_ALPHA = 0.10;
+
+    private final ConcurrentHashMap<InetAddressAndPort, AnnotatedEMA> samples = new ConcurrentHashMap<>();
+
+    /**
+     * Adds two boolean markers to the ExponentialMovingAverage for telling if the data has been
+     * updated or requested recently.
+     *
+     * recentlyMeasured is updated through {@link AnnotatedEMA#update(long, boolean)}
+     * recentlyRequested is updated through {@link DynamicEndpointSnitch#markRequested}
+     *
+     * Both markers are periodically reset via {@link DynamicEndpointSnitch#latencyProbeNeeded(long)}
+     */
+    private static class AnnotatedEMA extends ExponentialMovingAverage
+    {
+        volatile boolean recentlyRequested = false;
+        volatile boolean recentlyMeasured = false;
+
+        AnnotatedEMA(double alpha, double initialValue)
+        {
+            super(alpha, initialValue);
+        }
+
+        public void update(long value, boolean isRealRead)
+        {
+            recentlyMeasured = recentlyMeasured || isRealRead;
```

Comment:
This can be |=. This isn't checking of recently measured is already set before writing which means it's always going to pull the cache line in for write.

isRealRead sounds a little awkward and not proscriptive. What we are really trying to say is that it should count as recently measured. I think it might be clearer if it simply didn't override update.

I am also a developing fan of enums instead of booleans since true/false is just not that informative.

Reply:
You are right I did say boolean. The enum thing occurred to me later as it's usually a response to seeing a boolean in a method call and being "what does that do?" It's a new habit I am trying to build.